### PR TITLE
ED 209 update

### PR DESCRIPTION
**Pull request by Sindorman**
Depends on #5075 

- ED 209 uses verbal command like "stay, stop, patrol, arrest, detain".

- ED 209 can arrest/detain people if given the command.

### CODE REVIEW

**Comment by Sindorman** [comment]
!review

File: code/modules/mob/living/bot/ed209bot.dm [hunk 1]
```diff
@@ -19,6 +29,201 @@
...
+	check_records = FALSE
+	custom_emote(2, "[emote_hear], \"Roger that, going into idle mode. Auto patrol disabled [target]\"")
+	return 1
+
+/mob/living/bot/secbot/ed209/proc/stop_command(var/mob/speaker,var/text)
+	if(!on)
+		return
+	walk_to(src, src, 0, move_to_delay)
+	check_records = FALSE
+	custom_emote(2, "[emote_hear], \"Roger that, unit going offline. [target]\"")
```
Comment (Arrow768) [inline code/modules/mob/living/bot/ed209bot.dm]: Same here. Stray [target]

File: code/modules/mob/living/bot/ed209bot.dm [hunk 2]
```diff
@@ -19,6 +29,201 @@
...
+	turn_off()
+	return 1
+
+/mob/living/bot/secbot/ed209/proc/patrol_command(var/mob/speaker,var/text)
+	walk_to(src, src, 0, move_to_delay)
+	mode = SECBOT_IDLE
+	auto_patrol = 1
+	target = null
+	custom_emote(2, "[emote_hear], \"Roger that, starting patrol now [target]\"")
```
Comment (Arrow768) [inline code/modules/mob/living/bot/ed209bot.dm]: Same here. Stray [target]

File: code/modules/mob/living/bot/ed209bot.dm [hunk 3]
```diff
@@ -19,6 +29,201 @@
...
+
+	return 1
+
+//returns a list of everybody we wanna do stuff with.
+/mob/living/bot/secbot/ed209/proc/get_target_by_name(var/text)
+	var/list/possible_targets = hearers(src,10)
+	for(var/mob/M in possible_targets)
+		var/found = 0
+		if(findtext(text, "[M]"))
+			found = 1
```
Comment (BurgerLUA) [inline code/modules/mob/living/bot/ed209bot.dm]: Standard coding practice is just ```return M``` if you found something

File: code/modules/mob/living/bot/ed209bot.dm [hunk 4]
```diff
@@ -19,6 +29,201 @@
...
+
+/mob/living/bot/secbot/ed209/verb/change_name()
+	set name = "Change nickname"
+	set category = "IC"
+	set src in view(1)
+
+	var/mob/M = usr
+	if(!M.mind)	return 0
+
+	var/short_input = sanitizeSafe(input("What do you want [src] respond to?", , ""), MAX_NAME_LEN)
```
Comment (BurgerLUA) [inline code/modules/mob/living/bot/ed209bot.dm]: Minor nitpick, but I don't see any restrictions for changing the nickname of the ed209 unit. Would a criminal be able to change the nickname of the ED-209 unit to something else?

File: code/modules/mob/living/bot/secbot.dm [hunk 5]
```diff
@@ -510,6 +517,112 @@
...
+			var/obj/item/weapon/card/id/id = H.GetIdCard()
+			if(id)
+				perpname = id.registered_name
+
+			var/datum/data/record/R = find_security_record("name", perpname)
+			if(R)
+				R.fields["criminal"] = "*Arrest*"
+			else
+				check_records = TRUE
+			broadcast_security_hud_message("[src] was shot with <b>[P]</b>, projectile came from <b>[target]</b>, [arrest_type ? "detaining" : "arresting"] a level [check_threat(target)] suspect in <b>[get_area(src)]</b>. Requesting backup", src)
```
Comment (BurgerLUA) [inline code/modules/mob/living/bot/secbot.dm]: Another minor nitpick, there may be issues with super long range combat. For example, a sniper could hit the ed-209 unit and automatically identify someone despite being a long ways away.

File: code/modules/mob/living/bot/ed209bot.dm (hunk 4 above)
Comment (Sindorman) [reply code/modules/mob/living/bot/ed209bot.dm]: That is a good point

File: code/modules/mob/living/bot/secbot.dm (hunk 5 above)
Comment (Sindorman) [reply code/modules/mob/living/bot/secbot.dm]: That is a good point

File: code/modules/mob/living/bot/ed209bot.dm (hunk 3 above)
Comment (Sindorman) [reply code/modules/mob/living/bot/ed209bot.dm]: Yeah, you are right. That is a weird way. I copied code from commanded mobs and did not revise it entirely.

Comment (BurgerLUA) [reply code/modules/mob/living/bot/ed209bot.dm]: It's fine. I copy existing code all the time and usually it ends up being bad code.

File: code/modules/mob/living/bot/secbot.dm (hunk 5 above)
Comment (Sindorman) [reply code/modules/mob/living/bot/secbot.dm]: Done, I made a check up there that checks if we can see mob that shot us.

File: code/modules/mob/living/bot/ed209bot.dm [hunk 6]
```diff
@@ -19,14 +29,222 @@
...
+		if(id)
+			perpname = id.registered_name
+
+		var/datum/data/record/R = find_security_record("name", perpname)
+		if(R && R.fields["criminal"] == "*Arrest*")
+			visible_message(usr, "<span class='warning'>Warning, you do not have access!</span>")
+
+	// Check if they have access to the bot
+	if(!has_ui_access(usr))
+		visible_message(usr, "<span class='warning'>Warning, you do not have access!</span>")
```
Comment (skull132) [inline code/modules/mob/living/bot/ed209bot.dm]: Thaaat's not how `visible_message` woooorks. You want `to_chat(usr, "bla")`. `visible_message()` generates a visible message for everyone in the surrounding area to see.

File: code/modules/mob/living/bot/ed209bot.dm [hunk 7]
```diff
@@ -19,14 +29,222 @@
...
+	if(ishuman(usr))
+		var/mob/living/carbon/human/H = usr
+		var/perpname = H.name
+		var/obj/item/weapon/card/id/id = H.GetIdCard()
+		if(id)
+			perpname = id.registered_name
+
+		var/datum/data/record/R = find_security_record("name", perpname)
+		if(R && R.fields["criminal"] == "*Arrest*")
+			visible_message(usr, "<span class='warning'>Warning, you do not have access!</span>")
```
Comment (skull132) [inline code/modules/mob/living/bot/ed209bot.dm]: `visible_message(usr, "bla")` -> `to_chat(usr, "bla")`

File: code/modules/mob/living/bot/ed209bot.dm [hunk 8]
```diff
@@ -19,14 +29,222 @@
...
+	custom_emote(2, "[emote_hear], \"Roger that, following [target]\"")
+
+	return 1
+
+/mob/living/bot/secbot/ed209/verb/change_name()
+	set name = "Change nickname"
+	set category = "IC"
+	set src in view(1)
+
+	var/mob/M = usr
```
Comment (skull132) [inline code/modules/mob/living/bot/ed209bot.dm]: `usr` should already be typed to `mob`. So you shouldn't need this, me thinks.

File: code/modules/mob/living/bot/ed209bot.dm [hunk 9]
```diff
@@ -1,9 +1,19 @@
+#define SECBOT_IDLE 		0		// idle
```
Comment (skull132) [inline code/modules/mob/living/bot/ed209bot.dm]: `#undef` all of these at the end of the file unless they're used elsewhere.

File: code/modules/mob/living/bot/ed209bot.dm (hunk 6 above)
Comment (Sindorman) [reply code/modules/mob/living/bot/ed209bot.dm]: I confused them

**Pull request merged**